File: python/problems/leetcode/0092_reverse_linked_list_ii/solution.py
"""LeetCode 92: Reverse Linked List II."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def reverseBetween(
        self, head: Optional[ListNode], left: int, right: int
    ) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        pre = dummy
        for _ in range(left - 1):
            assert pre.next is not None
            pre = pre.next
        cur = pre.next
        for _ in range(right - left):
            assert cur is not None and cur.next is not None
            nxt = cur.next
            cur.next = nxt.next
            nxt.next = pre.next
            pre.next = nxt
        return dummy.next


def from_list(vs: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur = dummy
    for v in vs:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(h: Optional[ListNode]) -> list[int]:
    out: list[int] = []
    while h is not None:
        out.append(h.val)
        h = h.next
    return out


if __name__ == "__main__":
    assert to_list(Solution().reverseBetween(from_list([1, 2, 3, 4, 5]), 2, 4)) == [
        1, 4, 3, 2, 5,
    ]
    assert to_list(Solution().reverseBetween(from_list([5]), 1, 1)) == [5]
    print("0092 OK")

File: cpp/problems/leetcode/0092_reverse_linked_list_ii/solution.cpp
// LeetCode 92
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int v) : val(v), next(nullptr) {}
    ListNode(int v, ListNode* n) : val(v), next(n) {}
};

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode dummy(0, head);
        ListNode* pre = &dummy;
        for (int i = 0; i < left - 1; ++i) pre = pre->next;
        ListNode* cur = pre->next;
        for (int i = 0; i < right - left; ++i) {
            ListNode* nxt = cur->next;
            cur->next = nxt->next;
            nxt->next = pre->next;
            pre->next = nxt;
        }
        return dummy.next;
    }
};

static ListNode* fromList(vector<int> vs) {
    ListNode dummy;
    ListNode* cur = &dummy;
    for (int v : vs) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

static vector<int> toList(ListNode* h) {
    vector<int> out;
    while (h) {
        out.push_back(h->val);
        h = h->next;
    }
    return out;
}

int main() {
    assert((toList(Solution().reverseBetween(fromList({1, 2, 3, 4, 5}), 2, 4)) ==
            vector<int>{1, 4, 3, 2, 5}));
    assert((toList(Solution().reverseBetween(fromList({5}), 1, 1)) ==
            vector<int>{5}));
    cout << "0092 OK" << endl;
    return 0;
}
